vignettes/preprocess.Rmd

```diff
@@ -0,0 +1,346 @@
+---
+title: "preprocess"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{preprocess}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+Before running machine learning, it's often necessary and prudent to preprocess your input data. 
+We provide a function (`preprocess_data`) to preprocess input data. 
+The defaults we chose are based on best practices used in  [FIDDLE](https://gitlab.eecs.umich.edu/mld3/FIDDLE/-/tree/master/).
+Feel free to check out FIDDLE for more information about data preprocessing!
+
+The `preprocess_data` function takes an input dataset where the rows are the samples and the columns are the outcome variable and features.
+We preprocess the data as follows:
+
+- Remove missing outcome values. 
+- Leave binary features as-is (except that categorical variables are converted to 0 and 1, and binary variables with missing features are split into two rows - see below for more details).
+- Normalize continuous features using `caret::preProcess` based on the method provided.
+- Convert categorical features with more than 2 variables to 0 and 1 in multiple columns (one for each variable - not full rank, so each variable has it's own column).
+- Replace missing categorical data with 0.
+- Impute missing continuous values with the median of the feature. 
+- By default, remove all features with near-zero variance.
+- By default, collapse correlated features.
+
+# Examples
+
+We're going to start off simple and get more complicated, but if you want the whole shebang at once, just scroll to the bottom.
+
+First, we have to load `mikropml`:
+
+```{r setup}
+library(mikropml)
+```
+
+## Binary data
+
+Let's start with only binary variables:
+
+```{r}
+# raw binary dataset
+bin_df <- data.frame(
+  outcome = c("normal", "normal", "cancer"),
+  var1 = c("no", "yes", "no"),
+  var2 = c(0, 1, 1)
+)
+bin_df
+```
+
+In addition to the dataframe itself, you have to provide the name of the outcome column to `preprocess_data`. Here's what the preprocessed data looks like:
+
+```{r}
+# preprocess raw binary data
+preprocess_data(dataset = bin_df, outcome_colname = "outcome")
+```
+
+The output is a list of length two: `dat_transformed` which has the transformed data, and `grp_feats` which is a list of grouped features. Here, `grp_feats` is `NULL` because there are no perfectly correlated features (e.g. `c(0,1,0)` and `c(0,1,0)`, or `c(0,1,0)` and `c(1,0,1)` - see below for more details). 
+
+The first column (`var1`) in `dat_transformed` is changed to `var1_yes` that has zeros (no) and ones (yes). The values in the second column (`var2`) stay the same because it's already binary, but the name changes to `var2_1`. 
```

Comment:
Is var2 not a factor here then? If it was a factor, would we also split var2 to 2 columns?

Reply:
Good point! I had never thought about what would happen with factors. It turns out that it's treated the same way as a character vector. I added an example using a factor to the binary data and just pushed the changes.